File: apng/src/main/java/com/github/penfeizhou/animation/apng/decode/APNGDecoder.java
package com.github.penfeizhou.animation.apng.decode;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.Rect;

import com.github.penfeizhou.animation.apng.io.APNGReader;
import com.github.penfeizhou.animation.apng.io.APNGWriter;
import com.github.penfeizhou.animation.decode.Frame;
import com.github.penfeizhou.animation.decode.FrameSeqDecoder;
import com.github.penfeizhou.animation.io.Reader;
import com.github.penfeizhou.animation.loader.Loader;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

/**
 * @Description: APNG4Android
 * @Author: pengfei.zhou
 * @CreateDate: 2019-05-13
 */
public class APNGDecoder extends FrameSeqDecoder<APNGReader, APNGWriter> {

    private APNGWriter apngWriter;
    private int mLoopCount;
    private final Paint paint = new Paint();


    private static class SnapShot {
        byte dispose_op;
        Rect dstRect = new Rect();
        ByteBuffer byteBuffer;
    }

    private final SnapShot snapShot = new SnapShot();

    /**
     * @param loader         webp的reader
     * @param renderListener 渲染的回调
     */
    public APNGDecoder(Loader loader, RenderListener renderListener) {
        super(loader, renderListener);
        paint.setAntiAlias(true);
    }

    @Override
    protected APNGWriter getWriter() {
        if (apngWriter == null) {
            apngWriter = new APNGWriter();
        }
        return apngWriter;
    }

    @Override
    protected APNGReader getReader(Reader reader) {
        return new APNGReader(reader);
    }

    @Override
    protected int getLoopCount() {
        return mLoopCount;
    }

    @Override
    protected void release() {
        snapShot.byteBuffer = null;
        apngWriter = null;
    }


    @Override
    protected Rect read(APNGReader reader) throws IOException {
        List<Chunk> chunks = APNGParser.parse(reader);
        List<Chunk> otherChunks = new ArrayList<>();

        boolean actl = false;
        APNGFrame lastFrame = null;
        byte[] ihdrData = new byte[0];
        int canvasWidth = 0, canvasHeight = 0;
        for (Chunk chunk : chunks) {
            if (chunk instanceof ACTLChunk) {
                mLoopCount = ((ACTLChunk) chunk).num_plays;
                actl = true;
            } else if (chunk instanceof FCTLChunk) {
                APNGFrame frame = new APNGFrame(reader, (FCTLChunk) chunk);
                frame.prefixChunks = otherChunks;
                frame.ihdrData = ihdrData;
                frames.add(frame);
                lastFrame = frame;
            } else if (chunk instanceof FDATChunk) {
                if (lastFrame != null) {
                    lastFrame.imageChunks.add(chunk);
                }
            } else if (chunk instanceof IDATChunk) {
                if (!actl) {
                    // If it is a non-APNG image, only PNG will be decoded
                    Frame<APNGReader, APNGWriter> frame = new StillFrame(reader);
                    frame.frameWidth = canvasWidth;
                    frame.frameHeight = canvasHeight;
                    frames.add(frame);
                    mLoopCount = 1;
                    break;
                }
                if (lastFrame != null) {
                    lastFrame.imageChunks.add(chunk);
                }

            } else if (chunk instanceof IHDRChunk) {
                canvasWidth = ((IHDRChunk) chunk).width;
                canvasHeight = ((IHDRChunk) chunk).height;
                ihdrData = ((IHDRChunk) chunk).data;
            } else if (!(chunk instanceof IENDChunk)) {
                otherChunks.add(chunk);
            }
        }
        frameBuffer = ByteBuffer.allocate((canvasWidth * canvasHeight / (sampleSize * sampleSize) + 1) * 4);
        snapShot.byteBuffer = ByteBuffer.allocate((canvasWidth * canvasHeight / (sampleSize * sampleSize) + 1) * 4);
        return new Rect(0, 0, canvasWidth, canvasHeight);
    }

    @Override
    protected void renderFrame(Frame<APNGReader, APNGWriter> frame) {
        if (frame == null || fullRect == null) {
            return;
        }
        try {
            Bitmap bitmap = obtainBitmap(fullRect.width() / sampleSize, fullRect.height() / sampleSize);
            Canvas canvas = cachedCanvas.get(bitmap);
            if (canvas == null) {
                canvas = new Canvas(bitmap);
                cachedCanvas.put(bitmap, canvas);
            }
            if (frame instanceof APNGFrame) {
                // Restore the current frame from the cache
                frameBuffer.rewind();
                bitmap.copyPixelsFromBuffer(frameBuffer);
                // Process the settings in the snapshot before starting to draw
                if (this.frameIndex == 0) {
                    canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR);
                } else {
                    canvas.save();
                    canvas.clipRect(snapShot.dstRect);
                    switch (snapShot.dispose_op) {
                        // Restore the display from the snapshot before the last frame
                        case FCTLChunk.APNG_DISPOSE_OP_PREVIOUS:
                            snapShot.byteBuffer.rewind();
                            bitmap.copyPixelsFromBuffer(snapShot.byteBuffer);
                            break;
                        // Clear the area drawn in the previous frame
                        case FCTLChunk.APNG_DISPOSE_OP_BACKGROUND:
                            canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR);
                            break;
                        // Do nothing
                        case FCTLChunk.APNG_DISPOSE_OP_NON:
                        default:
                            break;
                    }
                    canvas.restore();
                }

                // Then pass it to the snapshot information according to the dispose setting
                if (((APNGFrame) frame).dispose_op == FCTLChunk.APNG_DISPOSE_OP_PREVIOUS) {
                    if (snapShot.dispose_op != FCTLChunk.APNG_DISPOSE_OP_PREVIOUS) {
                        snapShot.byteBuffer.rewind();
                        bitmap.copyPixelsToBuffer(snapShot.byteBuffer);
                    }
                }

                snapShot.dispose_op = ((APNGFrame) frame).dispose_op;
                canvas.save();
                if (((APNGFrame) frame).blend_op == FCTLChunk.APNG_BLEND_OP_SOURCE) {
                    canvas.clipRect(
                            frame.frameX / sampleSize,
                            frame.frameY / sampleSize,
                            (frame.frameX + frame.frameWidth) / sampleSize,
                            (frame.frameY + frame.frameHeight) / sampleSize);
                    canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR);
                }


                snapShot.dstRect.set(frame.frameX / sampleSize,
                        frame.frameY / sampleSize,
                        (frame.frameX + frame.frameWidth) / sampleSize,
                        (frame.frameY + frame.frameHeight) / sampleSize);
                canvas.restore();
            }
            // Start actually drawing the content of the current frame
            Bitmap inBitmap = obtainBitmap(frame.frameWidth, frame.frameHeight);
            recycleBitmap(frame.draw(canvas, paint, sampleSize, inBitmap, getWriter()));
            recycleBitmap(inBitmap);
            frameBuffer.rewind();
            bitmap.copyPixelsToBuffer(frameBuffer);
            recycleBitmap(bitmap);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: apng/src/main/java/com/github/penfeizhou/animation/apng/decode/APNGDecoder.kt
package com.github.penfeizhou.animation.apng.decode

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.Rect
import android.util.Size
import com.github.penfeizhou.animation.decode.FrameSeqDecoder2
import com.github.penfeizhou.animation.decode.ImageInfo
import com.github.penfeizhou.animation.decode.Frame
import com.github.penfeizhou.animation.decode.area
import com.github.penfeizhou.animation.io.ByteBufferWriter
import com.github.penfeizhou.animation.io.FilterReader
import com.github.penfeizhou.animation.io.Writer
import com.github.penfeizhou.animation.loader.Loader
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * @param loader webp-like reader
 *
 * @Description: APNG4Android
 * @Author: pengfei.zhou
 * @CreateDate: 2019-05-13
 */
class APNGDecoder(loader: Loader) : FrameSeqDecoder2(loader) {
    private val paint = Paint().apply { isAntiAlias = true }

    private class SnapShot {
        var disposeOp: Byte = 0
        val dstRect = Rect()
        var byteBuffer: ByteBuffer? = null
    }

    private val snapShot = SnapShot()

    private val apngWriter: Writer by lazy { ByteBufferWriter(ByteOrder.BIG_ENDIAN) }

    override fun release() {
        snapShot.byteBuffer = null
    }

    @Throws(IOException::class)
    override fun read(reader: FilterReader, sampleSize: Int): ImageInfo {
        val result = APNGParser.parse(reader)

        val isAnimated = result.actlChunk != null
        val loopCount = result.actlChunk?.num_plays ?: 1
        val viewport = Size(result.ihdrChunk.width, result.ihdrChunk.height)
        val frames = mutableListOf<Frame>()
        when {
            isAnimated ->
                result.frameDatas.mapIndexedTo(frames) { index, frameData ->
                    APNGFrame(
                        index,
                        reader,
                        frameData.fctlChunk,
                        result.ihdrChunk.data,
                        result.prefixChunks,
                        frameData.imageChunks
                    )
                }

            result.hasIDATChunk ->
                // If it is a non-APNG image, only PNG will be decoded
                frames += StillFrame(reader, viewport.width, viewport.height)
        }

        val bufferSizeBytes = (viewport.area / (sampleSize * sampleSize) + 1) * 4
        snapShot.byteBuffer = ByteBuffer.allocate(bufferSizeBytes)
        return ImageInfo(loopCount, viewport, frames)
    }

    override fun renderFrame(
        imageInfo: ImageInfo,
        frame: Frame,
        frameBuffer: ByteBuffer,
        sampleSize: Int
    ) {
        try {
            val bitmap = obtainBitmap(
                width = imageInfo.viewport.width / sampleSize,
                height = imageInfo.viewport.height / sampleSize
            ) ?: return
            val canvas = getCanvas(bitmap)

            if (frame is APNGFrame) {
                prepareApngBitmap(frame, bitmap, canvas, frameBuffer, sampleSize)
            }
            // Start actually drawing the content of the current frame
            val inBitmap = obtainBitmap(frame.width, frame.height) ?: return
            recycleBitmap(frame.draw(canvas, paint, sampleSize, inBitmap, apngWriter))
            recycleBitmap(inBitmap)
            frameBuffer.rewind()
            bitmap.copyPixelsToBuffer(frameBuffer)
            recycleBitmap(bitmap)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    // TODO: Not sure this is a suitable name
    private fun prepareApngBitmap(
        frame: APNGFrame,
        bitmap: Bitmap,
        canvas: Canvas,
        frameBuffer: ByteBuffer,
        sampleSize: Int
    ) {
        // Restore the current frame from the cache
        frameBuffer.rewind()
        bitmap.copyPixelsFromBuffer(frameBuffer)
        // Process the settings in the snapshot before starting to draw
        if (frame.index == 0) {
            canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR)
        } else {
            canvas.save()
            canvas.clipRect(snapShot.dstRect)
            when (snapShot.disposeOp) {
                FCTLChunk.APNG_DISPOSE_OP_PREVIOUS -> {
                    snapShot.byteBuffer?.rewind()
                    bitmap.copyPixelsFromBuffer(snapShot.byteBuffer)
                }
                FCTLChunk.APNG_DISPOSE_OP_BACKGROUND ->
                    canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR)
                FCTLChunk.APNG_DISPOSE_OP_NON -> Unit
            }
            canvas.restore()
        }

        // Then pass it to the snapshot information according to the dispose setting
        if (frame.disposeOp == FCTLChunk.APNG_DISPOSE_OP_PREVIOUS) {
            if (snapShot.disposeOp != FCTLChunk.APNG_DISPOSE_OP_PREVIOUS) {
                snapShot.byteBuffer?.rewind()
                bitmap.copyPixelsToBuffer(snapShot.byteBuffer)
            }
        }
        snapShot.disposeOp = frame.disposeOp
        canvas.save()
        if (frame.blendOp == FCTLChunk.APNG_BLEND_OP_SOURCE) {
            canvas.clipRect(
                frame.x / sampleSize,
                frame.y / sampleSize,
                (frame.x + frame.width) / sampleSize,
                (frame.y + frame.height) / sampleSize
            )
            canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR)
        }
        snapShot.dstRect.set(
            frame.x / sampleSize,
            frame.y / sampleSize,
            (frame.x + frame.width) / sampleSize,
            (frame.y + frame.height) / sampleSize
        )
        canvas.restore()
    }
}
